include/envoy/server/listener_manager.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "envoy/json/json_object.h"
+#include "envoy/network/filter.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/ssl/context.h"
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Factory for creating listen sockets.
+ */
+class ListenSocketFactory {
+public:
+  virtual ~ListenSocketFactory() {}
+
+  /**
+   * Creates a socket.
+   * @param address supplies the socket's address.
+   * @param bind_to_port supplies whether to actually bind the socket.
+   * @return Network::ListenSocketPtr an initialized and potentially bound socket.
+   */
+  virtual Network::ListenSocketPtr create(Network::Address::InstanceConstSharedPtr address,
+                                          bool bind_to_port) PURE;
+};
+
+/**
+ * A configuration for an individual listener.
+ */
+class Listener {
+public:
+  virtual ~Listener() {}
+
+  /**
+   * @return Network::FilterChainFactory& the factory for setting up the filter chain on a new
+   *         connection.
+   */
+  virtual Network::FilterChainFactory& filterChainFactory() PURE;
+
+  /**
+   * @return Network::Address::InstanceConstSharedPtr the address.
```

Comment:
"the configured address for the listener. This may be distinct to the bound address, e.g. if the port is zero."